tor/controller.go

```diff
@@ -169,6 +177,56 @@ func (c *Controller) Stop() error {
 	return c.conn.Close()
 }
 
+// Reconnect makes a new socket connection between the tor controller and
+// daemon. It will attempt to close the old connection, make a new connection
+// and authenticate, and finally delete the old Onion Service if there was one.
+func (c *Controller) Reconnect() error {
+	// Require the tor controller to be running when we want to reconnect.
+	// This means the started flag must be 1 and the stopped flag must be
+	// 0.
+	if c.started != 1 {
+		return errTCNotStarted
+	}
+	if c.stopped != 0 {
+		return errTCStopped
+	}
+
+	log.Info("Re-connectting tor controller")
+
+	// If we have an old connection, try to close it. We might receive an
+	// error if the connection has already been closed by Tor daemon(ie,
+	// daemon restarted), so we ignore the error here.
+	if c.conn != nil {
+		if err := c.conn.Close(); err != nil {
+			log.Debugf("closing old conn got err: %v", err)
+		}
+	}
+
+	// Make a new connection and authenticate.
+	conn, err := textproto.Dial("tcp", c.controlAddr)
+	if err != nil {
+		return fmt.Errorf("unable to connect to Tor server: %w", err)
+	}
+
+	c.conn = conn
+	if err := c.authenticate(); err != nil {
+		return err
+	}
+
+	// Exit early if we have no activeServiceID set.
+	if c.activeServiceID == "" {
+		return nil
+	}
+
+	// Remove the old Onion Service. The service might not be found once
+	// the Tor daemon restarted, so we ignore the error here.
+	if err := c.DelOnion(c.activeServiceID); err != nil {
```

Comment:
nit: add a comment here that `activeServiceID` is still the old service? Had to go read `authenticate` to check that it doesn't update this value + above maybe clarify that it would only be set for old services?

Reply:
Updated!